src/state/haproxy_route_tcp.py

```diff
@@ -0,0 +1,134 @@
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""HAproxy route charm state component."""
+
+from functools import cached_property
+from typing import Optional, Self, cast
+
+from charms.haproxy.v0.haproxy_route_tcp import (
+    HaproxyRouteTcpRequirerData,
+    TCPHealthCheckType,
+    TCPServerHealthCheck,
+)
+from pydantic import IPvAnyAddress, RootModel
+from pydantic.dataclasses import dataclass
+
+
+@dataclass(frozen=True)
+class HaproxyRouteTcpServer:
+    """A representation of a server in the backend section of the haproxy config.
+
+    Attrs:
+        server_name: The name of the unit with invalid characters replaced.
+        address: The IP address of the requirer unit.
+        port: The port that the requirer application wishes to be exposed.
+        check: Health check configuration.
+        maxconn: Maximum allowed connections before requests are queued.
+    """
+
+    server_name: str
+    address: IPvAnyAddress
+    port: int
+    check: Optional[TCPServerHealthCheck]
+    maxconn: Optional[int]
+
+
+@dataclass
+class HAProxyRouteTcpEndpoint(HaproxyRouteTcpRequirerData):
+    """Represent an endpoint for haproxy-route-tcp.
+
+    Attrs:
+        consistent_hashing: Whether consistent hashing should be applied for this backend.
+        servers: List of backend servers for this TCP endpoint.
+        name: Unique name for this TCP endpoint.
+        tcp_check_options: TCP health check options for HAProxy configuration.
+    """
+
+    @classmethod
+    def from_haproxy_route_tcp_requirer_data(cls, provider: HaproxyRouteTcpRequirerData) -> "Self":
+        """Instantiate a HAProxyRouteTcpEndpoint class from the parent class.
+
+        Args:
+            provider: parent class.
+
+        Returns:
+            Self: The instantiated HAProxyRouteTcpEndpoint class.
+        """
+        return cls(**RootModel[HaproxyRouteTcpRequirerData](provider).model_dump())
```

Comment:
I dont't like much the idea of deserializing with Pydantic, I feel we're introducing a point of dependency that we shouldn't need to

Reply:
We are depending on pydantic for validation, but do we need to depend on it for serialization too? Why not using the json package instead, for instance, which is part of the stdlib?

Reply:
yes, the `HAProxyRouteTcpEndpoint` is basically a subclass of `HaproxyRouteTcpRequirerData` that do additional processing on the relation data to add haproxy-specific fields to make rendering the template easier, while the parent class (`HaproxyRouteTcpRequirerData`) is specific to the library and should not include haproxy-specific language. Can you let me know what would you prefer if we're not using a subclass?

About initialization/serialization I'm only looking to have a dict of attributes of the parent class to use them to initialize the subclass to avoid having to also change this method when adding new attributes, I'll update it to explicitly specify the attributes to avoid the confusion.

Reply:
updated